dpctl/tensor/_dlpack.pyx

```diff
@@ -0,0 +1,399 @@
+
+cimport dpctl as c_dpctl
+cimport dpctl.memory as c_dpmem
+
+from .._backend cimport (
+    DPCTLDevice_Delete,
+    DPCTLDevice_GetParentDevice,
+    DPCTLSyclDeviceRef,
+    DPCTLSyclUSMRef,
+)
+from ._usmarray cimport usm_ndarray
+
+import numpy as np
+
+import dpctl
+import dpctl.memory as dpmem
+
+
+cdef extern from './include/dlpack/dlpack.h' nogil:
+    cdef int DLPACK_VERSION
+
+    cdef enum DLDeviceType:
+        kDLCPU
+        kDLCUDA
+        kDLCUDAHost
+        kDLCUDAManaged
+        kDLROCM
+        kDLROCMHost
+        kDLOpenCL
+        kDLVulkan
+        kDLMetal
+        kDLVPI
+        kDLOneAPI
+
+    ctypedef struct DLDevice:
+        DLDeviceType device_type
+        int device_id
+
+    cdef enum DLDataTypeCode:
+        kDLInt
+        kDLUInt
+        kDLFloat
+        kDLBfloat
+        kDLComplex
+
+    ctypedef struct DLDataType:
+        uint8_t code
+        uint8_t bits
+        uint16_t lanes
+
+    ctypedef struct DLTensor:
+        void* data
+        DLDevice device
+        int ndim
+        DLDataType dtype
+        int64_t* shape
+        int64_t* strides
+        uint64_t byte_offset
+
+    ctypedef struct DLManagedTensor:
+        DLTensor dl_tensor
+        void* manager_ctx
+        void (*deleter)(DLManagedTensor*)  # noqa: E211
+
+
+def get_build_dlpack_version():
+    """
+    Returns the string value of DLPACK_VERSION from dlpack.h
+    `dpctl.tensor` was built with.
+    """
+    return str(DLPACK_VERSION)
+
+
+cdef void _pycapsule_deleter(object dlt_capsule):
+    cdef DLManagedTensor *dlm_tensor = NULL
+    if cpython.PyCapsule_IsValid(dlt_capsule, 'dltensor'):
+        dlm_tensor = <DLManagedTensor*>cpython.PyCapsule_GetPointer(
+            dlt_capsule, 'dltensor')
+        dlm_tensor.deleter(dlm_tensor)
+
+
+cdef void _managed_tensor_deleter(DLManagedTensor *dlm_tensor) with gil:
+    if dlm_tensor is not NULL:
+        stdlib.free(dlm_tensor.dl_tensor.shape)
+        cpython.Py_DECREF(<usm_ndarray>dlm_tensor.manager_ctx)
+        dlm_tensor.manager_ctx = NULL
+        stdlib.free(dlm_tensor)
+
+
+cpdef to_dlpack_capsule(usm_ndarray usm_ary) except+:
+    """
+    Constructs named Python capsule object referencing
+    instance of `DLManagerTensor` from
+    :class:`dpctl.tensor.usm_ndarray` instance.
+    """
+    cdef c_dpctl.SyclQueue ary_sycl_queue
+    cdef c_dpctl.SyclDevice ary_sycl_device
+    cdef DPCTLSyclDeviceRef pDRef = NULL
+    cdef DLManagedTensor *dlm_tensor = NULL
+    cdef DLTensor* dl_tensor = NULL
+    cdef int nd = usm_ary.get_ndim()
+    cdef char* data_ptr = usm_ary.get_data()
+    cdef Py_ssize_t *shape_ptr = NULL
+    cdef Py_ssize_t *strides_ptr = NULL
+    cdef int64_t *shape_strides_ptr = NULL
+    cdef int i = 0
+    cdef int device_id = -1
+    cdef char* base_ptr = NULL
+    cdef Py_ssize_t element_offset = 0
+    cdef Py_ssize_t byte_offset = 0
+
+    ary_base = usm_ary.get_base()
+    ary_sycl_queue = usm_ary.get_sycl_queue()
```

Comment:
Are these copies or just references to the original pointer? If copies they should be freed at each excption handling block.

Reply:
These are Python objects.

Reply:
So garbage collector handles it for us.

Reply:
Ok disregard it in that case.